### PR TITLE
Table: Improvements to column resizing, style and alignment 

### PR DESCRIPTION
**Pull request by torkelo**
Fixes alignment issue for header cells and row cells when a native (non overlay) scrollbar for the rows was shown. This native scrollbar made all header rows not align perfectly with cells. To fix this I had to move to the absolute layout. 

* Switches to useAbsoluteLayout
* Adds a reducer so we can listen for internal react-table actions and persist resize state (now works pretty well)

Also from https://github.com/grafana/grafana/pull/23681

* [x] Switches to use data frame data directly, need to pass an array to react-table so passing it the first field array. Then using accessor function to get values. This way the react-table sort hook still works
* [x] Switches the column id to be the field index instead of the field name, so now the Table handles fields with the same name. 

Leaving sort by persisting to later PR as for that we need to decide if sorting:
* Should be done outside Table component or not
* Should we support multiple field sorting (works really well with current internal table sorting). Persisted as a simple array `[{id: 'columnId', desc: true}, {id: 'second-columnId', desc: false}]` 
* How should we persist sort state, we have multiple options
  * **A** As an internal table panel option (not even shown in UI, sort of like old angular panel). We can continue to use table panel internal sort if we want for now with this option. 
  * **B** As a field config override rule
  * **C** As a transformation (need to add function to change to PanelProps)

The problem with B & C is clean-up if you switch to another visualization the transform & potentially the override will remain. (And we have to implement multi-field sort if we want to support it but that should relatively easy)

### CODE REVIEW

File: packages/grafana-ui/src/components/CustomScrollbar/_CustomScrollbar.scss [hunk 1]
```diff
@@ -13,15 +13,15 @@
 
   .track-vertical {
     border-radius: 3px;
-    width: 6px !important;
+    width: 8px !important;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/CustomScrollbar/_CustomScrollbar.scss:16]: this is unrelated to main PR but good change, the custom scrollbars are to narrow

**Comment by ryantxu** [comment]
This is looking really great!  The one (ok I'm sure there is more than one!) thing that stands out is how to indicate that horizontal scrolling exists without a hover.  For example, in the dashboard view, this looks like a full table:

image

You have to hover over it to find out that the horizontal scrolling hides a bunch of columns:
image

Should we always show the horizontal scrollbar?  Maybe make it configurable?  (I vote we punt on this and add later if necessary)

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 2]
```diff
@@ -22,109 +31,142 @@ export interface Props {
...
-        minWidth: memoizedColumns.reduce((minWidth, column) => {
-          if (column.width) {
-            const width = typeof column.width === 'string' ? parseInt(column.width, 10) : column.width;
-            return Math.min(minWidth, width);
+interface ReactTableInternalState extends UseResizeColumnsState<{}>, UseSortByState<{}> {}
+
+function useTableStateReducer(props: Props) {
+  return useCallback(
+    (newState: ReactTableInternalState, action: any) => {
+      console.log(action, newState);
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Table/Table.tsx:43]: nit: maybe remove console.log

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -22,109 +31,142 @@ export interface Props {
...
+        case 'toggleSortBy':
+          if (props.onSortBy) {
+            // todo call callback and persist
+          }
+          break;
+      }
+
+      return newState;
+    },
+    [props.onColumnResize]
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Table/Table.tsx:63]:
```suggestion
    [props.onColumnResize, props.onSortBy]
```

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 4]
```diff
@@ -65,8 +97,8 @@ export class TablePanel extends Component<Props> {
 
     if (count > 1) {
```
Comment (hugohaggmark) [inline public/app/plugins/panel/table/TablePanel.tsx:98]: This means we're not rendering table for series with 1 row?
```suggestion
```

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 5]
```diff
@@ -13,21 +13,44 @@ export class TablePanel extends Component<Props> {
...
+    const propId = 'custom.width';
+
+    // look for existing override
+    const override = overrides.find(o => o.matcher.id === matcherId && o.matcher.options === fieldName);
+
+    if (override) {
+      // look for existing property
+      const property = override.properties.find(prop => prop.id === propId);
+      if (property) {
+        property.value = width;
```
Comment (hugohaggmark) [inline public/app/plugins/panel/table/TablePanel.tsx:41]: Can we mutate fieldConfigs like this?

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 6]
```diff
@@ -22,109 +31,142 @@ export interface Props {
...
+
+function useTableStateReducer(props: Props) {
+  return useCallback(
+    (newState: ReactTableInternalState, action: any) => {
+      console.log(action, newState);
+
+      switch (action.type) {
+        case 'columnDoneResizing':
+          if (props.onColumnResize) {
+            const info = (newState.columnResizing.headerIdWidths as any)[0];
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Table/Table.tsx:48]:
```suggestion
            const info:any = newState.columnResizing.headerIdWidths[0];
```

**Review by hugohaggmark — APPROVED** [review]
Great work and works as expected, I like the work you did on the Table. Cool that you can extend with a custom reducer. I left just some small comments for you to consider.

File: packages/grafana-ui/src/components/Table/Table.tsx (hunk 6 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/Table.tsx:48]: it's the react-table typings that are wrong that required me to do that

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 5 above)
Comment (torkelo) [reply public/app/plugins/panel/table/TablePanel.tsx:41]: we can, but would be better if done immutably.

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 4 above)
Comment (torkelo) [reply public/app/plugins/panel/table/TablePanel.tsx:98]: there render at the end of function for the other case of only 1 data frame. We now render data frame selector at bottom if TablePanel gets more than 1 data frame , this case if for that

File: packages/grafana-ui/src/components/Table/Table.tsx (hunk 2 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/Table.tsx:43]: going to remove this in later PR when sort by persistence is done

File: packages/grafana-ui/src/components/Table/Table.tsx (hunk 3 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/Table.tsx:63]: will fix when sort by persistence is added in this function

**Pull request merged**